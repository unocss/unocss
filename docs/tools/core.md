---
title: Core
description: The core engine of UnoCSS without any presets. It can be used as the engine of your own atomic CSS framework.
---

# Core

The core engine of UnoCSS without any presets: `@unocss/core`. It can be used as the engine of your own atomic CSS framework.

## Installation

::: code-group

```bash [pnpm]
pnpm add -D @unocss/core
```

```bash [yarn]
yarn add -D @unocss/core
```

```bash [npm]
npm install -D @unocss/core
```

```bash [bun]
bun add -D @unocss/core
```

:::

## Usage

```ts
import { createGenerator } from '@unocss/core'

const generator = await createGenerator(
  { /* user options */ },
  { /* default options */ }
)

const { css } = await generator.generate(code)
```

## License

- MIT License &copy; 2021-PRESENT [Anthony Fu](https://github.com/antfu)
