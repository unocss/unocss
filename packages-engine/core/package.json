{
  "name": "@unocss/core",
  "type": "module",
  "version": "66.1.0-beta.3",
  "description": "The instant on-demand Atomic CSS engine.",
  "author": "Anthony Fu <anthonyfu117@hotmail.com>",
  "license": "MIT",
  "funding": "https://github.com/sponsors/antfu",
  "homepage": "https://unocss.dev",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/unocss/unocss",
    "directory": "packages-engine/core"
  },
  "bugs": {
    "url": "https://github.com/unocss/unocss/issues"
  },
  "keywords": [
    "unocss",
    "atomic-css",
    "atomic-css-engine",
    "css",
    "tailwind",
    "windicss"
  ],
  "sideEffects": false,
  "exports": {
    ".": {
      "types": "./dist/index.d.mts",
      "default": "./dist/index.mjs"
    }
  },
  "main": "dist/index.mjs",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "unbuild",
    "stub": "unbuild --stub",
    "test:attw": "attw --pack --config-path ../../.attw-esm-only.json"
  },
  "devDependencies": {
    "magic-string": "catalog:",
    "unconfig": "catalog:"
  }
}
