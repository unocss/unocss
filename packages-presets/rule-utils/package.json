{
  "name": "@unocss/rule-utils",
  "type": "module",
  "version": "66.1.0-beta.3",
  "description": "Utilities for UnoCSS",
  "author": "Anthony Fu <anthonyfu117@hotmail.com>",
  "license": "MIT",
  "funding": "https://github.com/sponsors/antfu",
  "homepage": "https://unocss.dev",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/unocss/unocss",
    "directory": "packages-presets/rule-utils"
  },
  "bugs": {
    "url": "https://github.com/unocss/unocss/issues"
  },
  "keywords": [],
  "sideEffects": false,
  "exports": {
    ".": {
      "types": "./dist/index.d.mts",
      "default": "./dist/index.mjs"
    }
  },
  "main": "dist/index.mjs",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "engines": {
    "node": ">=14"
  },
  "scripts": {
    "build": "unbuild",
    "stub": "unbuild --stub",
    "test:attw": "attw --pack --config-path ../../.attw-esm-only.json"
  },
  "dependencies": {
    "@unocss/core": "workspace:^",
    "magic-string": "catalog:"
  }
}
