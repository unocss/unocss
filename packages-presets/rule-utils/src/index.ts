export * from './colors'
export * from './directive'
export * from './handlers'
export * from './icon'
export * from './utilities'
export * from './variants'
