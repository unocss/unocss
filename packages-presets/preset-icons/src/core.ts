import type { IconifyJSON } from '@iconify/types'
import type { IconifyLoaderOptions, UniversalIconLoader } from '@iconify/utils'
import type { CSSObject } from '@unocss/core'
import type { IconsOptions } from './types'
import { loadIcon } from '@iconify/utils/lib/loader/loader'
import { searchForIcon } from '@iconify/utils/lib/loader/modern'
import { encodeSvgForCss } from '@iconify/utils/lib/svg/encode-svg-for-css'
import { definePreset, warnOnce } from '@unocss/core'
import icons from './collections'

const COLLECTION_NAME_PARTS_MAX = 3

export { IconsOptions }
export { icons }

/**
 * API for preset-icons
 */
export interface IconsAPI {
  encodeSvgForCss: typeof encodeSvgForCss
  parseIconWithLoader: typeof parseIconWithLoader
  /**
   * This API only available for preset-icons created on Node.js environment
   */
  createNodeLoader?: () => Promise<UniversalIconLoader | undefined>
}

export function createPresetIcons(lookupIconLoader: (options: IconsOptions) => Promise<UniversalIconLoader>) {
  return definePreset((options: IconsOptions = {}) => {
    const {
      scale = 1,
      mode = 'auto',
      prefix = 'i-',
      warn = false,
      iconifyCollectionsNames,
      collections: customCollections,
      extraProperties = {},
      customizations = {},
      autoInstall = false,
      collectionsNodeResolvePath,
      layer = 'icons',
      unit,
      processor,
    } = options

    const flags = getEnvFlags()

    const loaderOptions: IconifyLoaderOptions = {
      addXmlNs: true,
      scale,
      customCollections,
      autoInstall,
      cwd: collectionsNodeResolvePath,
      // avoid warn from @iconify/loader: we'll warn below if not found
      warn: undefined,
      customizations: {
        ...customizations,
        additionalProps: { ...extraProperties },
        trimCustomSvg: true,
        async iconCustomizer(collection, icon, props) {
          await customizations.iconCustomizer?.(collection, icon, props)
          if (unit) {
            if (!props.width)
              props.width = `${scale}${unit}`
            if (!props.height)
              props.height = `${scale}${unit}`
          }
        },
      },
    }

    let iconLoader: UniversalIconLoader

    return {
      name: '@unocss/preset-icons',
      enforce: 'pre',
      options,
      layers: { icons: -30 },
      api: <IconsAPI>{
        encodeSvgForCss,
        parseIconWithLoader,
      },
      rules: [[
        /^([\w:-]+)(?:\?(mask|bg|auto))?$/,
        async (matcher) => {
          let [full, body, _mode = mode] = matcher as [string, string, IconsOptions['mode']]

          iconLoader = iconLoader || await lookupIconLoader(options)

          const usedProps = {}
          const parsed = await parseIconWithLoader(
            body,
            iconLoader,
            { ...loaderOptions, usedProps },
            iconifyCollectionsNames,
          )

          if (!parsed) {
            if (warn && !flags.isESLint)
              warnOnce(`failed to load icon "${full}"`)
            return
          }

          let cssObject: CSSObject
          const url = `url("data:image/svg+xml;utf8,${encodeSvgForCss(parsed.svg)}")`

          if (_mode === 'auto')
            _mode = parsed.svg.includes('currentColor') ? 'mask' : 'bg'

          if (_mode === 'mask') {
            // Thanks to https://codepen.io/noahblon/post/coloring-svgs-in-css-background-images
            cssObject = {
              '--un-icon': url,
              '-webkit-mask': 'var(--un-icon) no-repeat',
              'mask': 'var(--un-icon) no-repeat',
              '-webkit-mask-size': '100% 100%',
              'mask-size': '100% 100%',
              'background-color': 'currentColor',
              // for Safari https://github.com/elk-zone/elk/pull/264
              'color': 'inherit',
              ...usedProps,
            }
          }
          else {
            cssObject = {
              'background': `${url} no-repeat`,
              'background-size': '100% 100%',
              'background-color': 'transparent',
              ...usedProps,
            }
          }

          processor?.(cssObject, {
            ...parsed,
            icon: parsed.name,
            mode: _mode,
          })

          return cssObject
        },
        { layer, prefix },
      ]],
    }
  })
}

export function combineLoaders(loaders: UniversalIconLoader[]) {
  return (async (...args) => {
    for (const loader of loaders) {
      if (!loader)
        continue
      const result = await loader(...args)
      if (result)
        return result
    }
  }) as UniversalIconLoader
}

export function createCDNFetchLoader(
  fetcher: (url: string) => Promise<any>,
  cdnBase: string,
  cacheMap: Map<string, IconifyJSON | Promise<IconifyJSON>> = new Map(),
): UniversalIconLoader {
  function fetchCollection(name: string) {
    if (!icons.includes(name))
      return undefined
    if (!cacheMap.has(name))
      cacheMap.set(name, fetcher(`${cdnBase}@iconify-json/${name}/icons.json`))
    return cacheMap.get(name)!
  }

  return async (collection, icon, options) => {
    let result = await loadIcon(collection, icon, options)
    if (result)
      return result

    const iconSet = await fetchCollection(collection)
    if (iconSet) {
      // possible icon names
      const ids = [
        icon,
        icon.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase(),
        icon.replace(/([a-z])(\d+)/g, '$1-$2'),
      ]
      result = await searchForIcon(iconSet, collection, ids, options)
    }

    return result
  }
}

export function getEnvFlags() {
  // eslint-disable-next-line node/prefer-global/process
  const isNode = typeof process !== 'undefined' && process.stdout && !process.versions.deno
  // eslint-disable-next-line node/prefer-global/process
  const isVSCode = isNode && !!process.env.VSCODE_CWD
  // eslint-disable-next-line node/prefer-global/process
  const isESLint = isNode && !!process.env.ESLINT

  return {
    isNode,
    isVSCode,
    isESLint,
  }
}

export async function parseIconWithLoader(
  body: string,
  loader: UniversalIconLoader,
  options: IconifyLoaderOptions = {},
  safeCollectionsNames: string[] = [],
) {
  let collection = ''
  let name = ''
  let svg: string | undefined

  const allCollections = new Set<string>([
    ...icons,
    ...safeCollectionsNames,
    ...Object.keys(options.customCollections || {}),
  ])

  if (body.includes(':')) {
    [collection, name] = body.split(':')

    if (!allCollections.has(collection))
      return

    svg = await loader(collection, name, options)
  }
  else {
    const parts = body.split(/-/g)
    for (let i = COLLECTION_NAME_PARTS_MAX; i >= 1; i--) {
      collection = parts.slice(0, i).join('-')

      if (!allCollections.has(collection))
        continue

      name = parts.slice(i).join('-')
      svg = await loader(collection, name, options)

      if (svg)
        break
    }
  }

  if (!svg) {
    return
  }

  return {
    collection,
    name,
    svg,
  }
}
