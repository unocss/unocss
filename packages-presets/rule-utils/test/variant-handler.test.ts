import { createGenerator } from '@unocss/core'
import { variantMatcher } from '@unocss/rule-utils'
import { describe, expect, it } from 'vitest'

describe('variants', () => {
  it('variant context is propagated', async () => {
    const uno = await createGenerator({
      rules: [
        ['foo', { name: 'bar' }],
      ],
      variants: [
        {
          match(input) {
            const match = input.match(/^var:/)
            if (match) {
              return {
                matcher: input.slice(match[0].length),
                handle: (input, next) => next({
                  prefix: ':prefix > ',
                  selector: '.selector',
                  pseudo: '::pseudo',
                  entries: input.entries.map((entry) => {
                    entry[1] += ' !important'
                    return entry
                  }),
                  parent: '@supports',
                  layer: 'variant',
                }),
              }
            }
          },
        },
      ],
    })

    const { css } = await uno.generate([
      'foo',
      'var:foo',
    ].join(' '), { preflights: false })

    expect(css).toMatchSnapshot()
  })

  it('selector section is merged in order', async () => {
    const uno = await createGenerator({
      rules: [
        ['foo', { name: 'bar' }],
      ],
      variants: [
        variantMatcher('pre', () => ({ prefix: '.prefix ' })),
        variantMatcher('main', () => ({ selector: '.replaced' })),
        variantMatcher('back', () => ({ pseudo: '::pseudo' })),
      ],
    })

    const { css } = await uno.generate([
      'pre:main:foo',
      'pre:back:foo',
      'main:back:foo',
      'pre:main:back:foo',
      'back:main:pre:foo',
    ].join(' '), { preflights: false })

    expect(css).toMatchSnapshot()
  })

  it('variant can stack', async () => {
    const uno = await createGenerator({
      rules: [
        ['foo', { name: 'bar' }],
      ],
      variants: [
        {
          multiPass: true,
          match(input) {
            const match = input.match(/^(append-one|append-two|append-three):/)
            if (match) {
              return {
                matcher: input.slice(match[0].length),
                handle: (input, next) => next({
                  ...input,
                  selector: `${input.selector} > :${match[1]}`,
                }),
              }
            }
          },
        },
        {
          multiPass: true,
          match(input) {
            const match = input.match(/^(prepend-one|prepend-two|prepend-three):/)
            if (match) {
              return {
                matcher: input.slice(match[0].length),
                handle: (input, next) => {
                  const result = next(input)
                  return {
                    ...result,
                    selector: `${result.selector} + :${match[1]}`,
                  }
                },
              }
            }
          },
        },
      ],
    })

    const { css } = await uno.generate([
      'append-one:append-two:append-three:foo',
      'prepend-one:prepend-two:prepend-three:foo',
      'append-one:prepend-three:prepend-two:foo',
    ].join(' '), { preflights: false })

    expect(css).toMatchSnapshot()
  })

  it('noMerge on variant', async () => {
    const uno = await createGenerator({
      rules: [
        ['foo', { name: 'bar' }],
      ],
      variants: [
        {
          match(input) {
            const match = input.match(/^(var-([abc])):/)
            if (match) {
              return {
                matcher: input.slice(match[0].length),
                handle: (input, next) => next({
                  ...input,
                  pseudo: `${input.pseudo}::${match[2]}`,
                  noMerge: match[2] === 'c',
                }),
              }
            }
          },
        },
      ],
    })

    const { css } = await uno.generate([
      'foo',
      'var-a:foo',
      'var-b:foo',
      'var-c:foo',
    ].join(' '), { preflights: false })

    expect(css).toMatchSnapshot()
  })

  it('noMerge variant with shortcut', async () => {
    const uno = await createGenerator({
      rules: [
        ['foo', { name: 'bar' }],
      ],
      shortcuts: [
        ['bar', 'var-a:foo var-b:foo var-c:foo'],
      ],
      variants: [
        {
          match(input) {
            const match = input.match(/^(var-([abc])):/)
            if (match) {
              return {
                matcher: input.slice(match[0].length),
                handle: (input, next) => next({
                  ...input,
                  pseudo: `${input.pseudo}::${match[2]}`,
                  noMerge: match[2] === 'c',
                }),
              }
            }
          },
        },
      ],
    })

    const { css } = await uno.generate([
      'bar',
    ].join(' '), { preflights: false })

    expect(css).toMatchSnapshot()
  })

  it('selector match can be ordered', async () => {
    const uno = await createGenerator({
      rules: [
        ['foo', { name: 'bar' }],
      ],
      variants: [
        variantMatcher('pre', () => ({ selector: '.gone' })),
        {
          ...variantMatcher('pre', () => ({ selector: '.override' })),
          order: -1,
        },
      ],
    })

    const { css } = await uno.generate([
      'pre:foo',
    ].join(' '), { preflights: false })

    expect(css).toMatchSnapshot()
  })

  it('multiple handlers for single variant', async () => {
    const uno = await createGenerator({
      rules: [
        ['foo', { name: 'bar' }],
      ],
      variants: [
        variantMatcher('pre', [
          () => ({ prefix: '.prefix1 ' }),
          () => ({ prefix: '.prefix2 ' }),
        ]),
      ],
    })

    const { css } = await uno.generate([
      'pre:foo',
    ].join(' '), { preflights: false })

    expect(css).toMatchSnapshot()
  })
})
