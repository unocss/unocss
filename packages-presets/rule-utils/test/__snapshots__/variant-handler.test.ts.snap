// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`variants > multiple handlers for single variant 1`] = `
"/* layer: default */
.prefix1 .pre\\:foo,
.prefix2 .pre\\:foo{name:bar;}"
`;

exports[`variants > noMerge on variant 1`] = `
"/* layer: default */
.foo,
.var-a\\:foo::a,
.var-b\\:foo::b{name:bar;}
.var-c\\:foo::c{name:bar;}"
`;

exports[`variants > noMerge variant with shortcut 1`] = `
"/* layer: shortcuts */
.bar::a,
.bar::b{name:bar;}
.bar::c{name:bar;}"
`;

exports[`variants > selector match can be ordered 1`] = `
"/* layer: default */
.override{name:bar;}"
`;

exports[`variants > selector section is merged in order 1`] = `
"/* layer: default */
.prefix .pre\\:back\\:foo::pseudo,
.prefix .replaced,
.prefix .replaced::pseudo,
.replaced::pseudo{name:bar;}"
`;

exports[`variants > variant can stack 1`] = `
"/* layer: default */
.append-one\\:append-two\\:append-three\\:foo > :append-three > :append-two > :append-one,
.append-one\\:prepend-three\\:prepend-two\\:foo > :append-one + :prepend-three + :prepend-two,
.prepend-one\\:prepend-two\\:prepend-three\\:foo + :prepend-one + :prepend-two + :prepend-three{name:bar;}"
`;

exports[`variants > variant context is propagated 1`] = `
"/* layer: default */
.foo{name:bar;}
/* layer: variant */
@supports{
:prefix > .selector::pseudo{name:bar !important;}
}"
`;
